Unity/DynamicArray.cs

```diff
@@ -10,12 +13,31 @@ unsafe struct DynamicArray
         public ulong Capacity;
     }
 
-    unsafe struct DynamicArray<T>
+    unsafe struct DynamicArray<T> : IReadOnlyList<T>
         where T : unmanaged
     {
         public T* Ptr;
         public MemLabelId Label;
         public ulong Size;
         public ulong Capacity;
+
+        public T this[int index] => Ptr[index];
+
+        public int Count => (int)Size;
+
+        public IEnumerator<T> GetEnumerator()
+        {
+            var arr = new T[Count];
+            for(int i = 0; i < Count; i++)
+            {
+                arr[i] = Ptr[i];
+            }
+            return arr.AsEnumerable().GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }

```

Comment:
This should probably be replaced with a custom enumerator struct instead of copying everything to a managed array:
```cs
IEnumerator<T> IEnumerable<T>.GetEnumerator()
{
    return GetEnumerator();
}

IEnumerator IEnumerable.GetEnumerator()
{
    return GetEnumerator();
}

public Enumerator GetEnumerator()
{
    return new Enumerator(this);
}

public struct Enumerator : IEnumerator<T>
{
    T* current;

    readonly T* start;

    readonly T* end;

    public Enumerator(DynamicArray<T> array)
    {
        start   = array.Ptr;
        end     = array.Ptr + array.Size;
        current = start - 1;
    }

    public T Current => *current;

    object IEnumerator.Current => Current;

    public void Dispose()
    {
    }

    public bool MoveNext()
    {
        return ++current < end;
    }

    public void Reset()
    {
        current = start - 1;
    }
}
```

Reply:
Yeah, I went with a quick solution there. Enumerators are a bit of a pain to work with. I wish collections exposed Enumerable, then you could just yield and the compiler would handle the state machine for you.

Reply:
The quick solution is good enough for now, so I think I'll just merge and implement a custom enumerator later.
There's also the option of doing this if you want to `yield`:
```cs
public ref T this[int index] => ref Ptr[index];

public ref T this[ulong index] => ref Ptr[index];

public IEnumerator<T> GetEnumerator()
{
    for (ulong i = 0; i < Size; i++)
        yield return this[i];
}
```

Reply:
Oh, I had in my head you could only yield from methods that return an IEnumerable, if I knew you could yield from IEnumerator too I would have wrote what you suggested 
```csharp
public IEnumerator<T> GetEnumerator()
{
    for (ulong i = 0; i < Size; i++)
        yield return this[i];
}
```

Reply:
Yeah, both `IEnumerator` and `IEnumerable` can yield. I used to get confused all the time over when I could `yield` before I knew that.